### ISSUE TITLE
Java wrong java library version reported

### ISSUE THREAD

**Issue by pkroiss** [anchore/syft#3529]
We've a java project that uses jackson-databind-2.15.2.jar. Syft reports that we use 

jackson-databind                       2.11.1                                     java-archive  (+3 duplicates)  
jackson-databind                       2.15.2                                     java-archive 

We only use version 2.15.2

After investation we found in the ehcache-2.10.9.2.jar library 2 files where jackson-databind-2.11.1 is referenced
rest-management-private-classpath/META-INF/maven/com.fasterxml.jackson.core/jackson-databind/pom.xml
rest-management-private-classpath/META-INF/maven/com.fasterxml.jackson.core/jackson-databind/pom.properties
because of (+3 duplicates) we expect that similar META-INF entries exist in other libraries as well.

**Comment by spiffcs** [comment]
@pkroiss - since 2.11.1 is referenced in ehcache-2.10.9.2.jar is it correct to say it is a transitive dependency of your project?

Are you looking for an option where your SBOM only has dependencies that are "used" or delcared at the top level of your program?

**Comment by pkroiss** [comment]
Yes, I'm looking for an option that shows me the version that is actually used and in my case that is version 2.15.2

**Comment by spiffcs** [comment]
> Yes, I'm looking for an option that shows me the version that is actually used and in my case that is version 2.15.2

Sorry I don't want to misunderstand this before we investigate more.

Based on your comment I'm thinking syft's point of view your project technically uses both the `2.11.1` and `2.15.2` version just as different points of the dependency graph.
```
  1
 / \
2  3
   / \
  4  x

1: your project
2: jackson-databind-2.15.2
3: ehcache-2.10.9.2
4: jackson-databind-2.11.1
```

Do you have an example container I can scan or an SBOM to look at that shows what you think syft is doing that's incorrect here?

**Comment by kzantow** [comment]
Just to provide some further exploration: expanding the ehcache jar file results in LOTS of class files which have been shaded into the jar -- these are the files that are under `rest-management-private-classpath`, such as:
```
rest-management-private-classpath/com/fasterxml/jackson/databind/JsonDeserializer$None.class_terracotta
rest-management-private-classpath/com/fasterxml/jackson/databind/JsonDeserializer.class_terracotta
rest-management-private-classpath/com/fasterxml/jackson/databind/JsonMappingException$Reference.class_terracotta
rest-management-private-classpath/com/fasterxml/jackson/databind/JsonMappingException.class_terracotta
rest-management-private-classpath/com/fasterxml/jackson/databind/JsonNode$1.class_terracotta
rest-management-private-classpath/com/fasterxml/jackson/databind/JsonNode.class_terracotta
rest-management-private-classpath/com/fasterxml/jackson/databind/JsonSerializable$Base.class_terracotta
rest-management-private-classpath/com/fasterxml/jackson/databind/JsonSerializable.class_terracotta
```
So, it seems as though both versions are being used and syft is reporting the correct thing here. As such, I'm going to close this issue, but if this has been misunderstood, please let us know!

**Comment by pkroiss** [comment]
I've added a mvn project that demonstrate the problem.


unzip syft-demo.tar.gz
cd system-demo

./mvnw package

# versions used
grype --version
grype 0.87.0

syft --version
syft 1.19.0


syft scan target/spring-boot-rest-1.0.0.jar | grep jackson-databind
jackson-databind                 2.11.1            java-archive   -> wrong version reported not installed
jackson-databind                 2.12.3            java-archive   

grype target/spring-boot-rest-1.0.0.jar

NAME                       INSTALLED         FIXED-IN          TYPE          VULNERABILITY        SEVERITY 
jackson-databind           2.11.1            2.12.7.1          java-archive  GHSA-rgv9-q543-rqg4  High   -> wrong version reported not installed
jackson-databind           2.11.1            2.12.7.1          java-archive  GHSA-jjjh-jjxp-wpff  High   -> wrong version reported not installed
jackson-databind           2.11.1            2.12.6.1          java-archive  GHSA-57j2-w4cx-62h2  High   -> wrong version reported not installed    
jackson-databind           2.11.1            2.12.6            java-archive  GHSA-3x8x-79m2-3w2w  High   -> wrong version reported not installed
jackson-databind           2.12.3            2.12.7.1          java-archive  GHSA-rgv9-q543-rqg4  High      
jackson-databind           2.12.3            2.12.7.1          java-archive  GHSA-jjjh-jjxp-wpff  High      
jackson-databind           2.12.3            2.12.6.1          java-archive  GHSA-57j2-w4cx-62h2  High      
jackson-databind           2.12.3            2.12.6            java-archive  GHSA-3x8x-79m2-3w2w  High 


# Effective pom - 2.12.3 - should be used 
./mvnw help:effective-pom | grep -C3 jackson-databind
      </dependency>
      <dependency>
        <groupId>com.fasterxml.jackson.core</groupId>
        <artifactId>jackson-databind</artifactId>
        <version>2.12.3</version>
      </dependency>
      <dependency>



# find used version of jackson-databind 2.12.3 
cd target
unzip spring-boot-rest-1.0.0.jar

target % find . -name "*databind*.jar"
./BOOT-INF/lib/jackson-databind-2.12.3.jar


mkdir temp
cd temp 
BOOT-INF/lib/ehcache-2.10.9.2.jar temp/
cd temp
unzip ehcache-2.10.9.2.jar

# find version 2.11.1 - only a pom.xml - in a runtime
find . -name "pom.xml" | xargs grep -C3 jackson-databind

./rest-management-private-classpath/META-INF/maven/com.fasterxml.jackson.core/jackson-databind/pom.xml-  </parent>
./rest-management-private-classpath/META-INF/maven/com.fasterxml.jackson.core/jackson-databind/pom.xml-
./rest-management-private-classpath/META-INF/maven/com.fasterxml.jackson.core/jackson-databind/pom.xml-  <groupId>com.fasterxml.jackson.core</groupId>
./rest-management-private-classpath/META-INF/maven/com.fasterxml.jackson.core/jackson-databind/pom.xml:  <artifactId>jackson-databind</artifactId>
./rest-management-private-classpath/META-INF/maven/com.fasterxml.jackson.core/jackson-databind/pom.xml-  <version>2.11.1</version>
./rest-management-private-classpath/META-INF/maven/com.fasterxml.jackson.core/jackson-databind/pom.xml:  <name>jackson-databind</name>
./rest-management-private-classpath/META-INF/maven/com.fasterxml.jackson.core/jackson-databind/pom.xml-  <packaging>bundle</packaging>
./rest-management-private-classpath/META-INF/maven/com.fasterxml.jackson.core/jackson-databind/pom.xml-  <description>General data-binding functionality for Jackson: works on core streaming API</description>
./rest-management-private-classpath/META-INF/maven/com.fasterxml.jackson.core/jackson-databind/pom.xml-  <url>http://github.com/FasterXML/jackson</url>

[syft-demo.tar.gz](https://github.com/user-attachments/files/19209296/syft-demo.tar.gz)

**Comment by Kevin-Molina** [comment]
> Just to provide some further exploration: expanding the ehcache jar file results in LOTS of class files which have been shaded into the jar -- these are the files that are under `rest-management-private-classpath`, such as:
> 
> ```
> rest-management-private-classpath/com/fasterxml/jackson/databind/JsonDeserializer$None.class_terracotta
> rest-management-private-classpath/com/fasterxml/jackson/databind/JsonDeserializer.class_terracotta
> rest-management-private-classpath/com/fasterxml/jackson/databind/JsonMappingException$Reference.class_terracotta
> rest-management-private-classpath/com/fasterxml/jackson/databind/JsonMappingException.class_terracotta
> rest-management-private-classpath/com/fasterxml/jackson/databind/JsonNode$1.class_terracotta
> rest-management-private-classpath/com/fasterxml/jackson/databind/JsonNode.class_terracotta
> rest-management-private-classpath/com/fasterxml/jackson/databind/JsonSerializable$Base.class_terracotta
> rest-management-private-classpath/com/fasterxml/jackson/databind/JsonSerializable.class_terracotta
> ```
> 
> So, it seems as though both versions are being used and syft is reporting the correct thing here. As such, I'm going to close this issue, but if this has been misunderstood, please let us know!

@kzantow Is there a `syft` native way to show this that these are being shaded in by `ehcache` / owned by ehcache?

I know I can dump the syft output as json, then kind of infer it from `metadata.virtualPath` or `locations.accessPath` fields or similar, but is there any native functionality akin to `npm list` / various package manager hierarchy trees to report this info in a more user-friendly way?

Thank you!

**Comment by kzantow** [comment]
@Kevin-Molina sorry, I can't think of anything offhand that will help show this in any particularly more useful way. We _do_ output relationships, so you should be able to use that graph structure to understand where the relationships are, but none of Syft's native output will be much help visualizing the information any better.
